compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1855,7 +1838,7 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
     else if (!tp2.exists) tp1
     else tp.derivedAndType(tp1, tp2)
 
-  /** If some (&-operand of) `tp` is a supertype of `sub` replace it with `NoType`.
```

Comment:
This fix was unrelated to NotNull and should be kept.

Reply:
I reverted the reversion :)